Create DOCUMENTATION.md (#41)
ingokuba/swing-chat: DOCUMENTATION.md

# Advanced Software Enineering Project (DHBW Stuttgart)
## A Chat-Simulation 
### Made by Henrik Storch (2909251) and Ingo Kuba (3373309)

# Conent
### 1. [What it does](#1)
### 2. [Design](#design)
### 3. [Implementation](#Implementation)
### 4. [Design Patterns](#4)
### 5. [Sources](#5)

-----

# What it does <a name="1"><a/>

This project demonstrates a chat application on one local computer. Once the project is build and started, there is a so called "Main Window". It provides a method for new 'users' to join the chat. After inserting a name which is not yet taken by another user and confirmation, a new window will open, showing the chat client for the new user.

> Each client window has in the top left a list with all active users. The user of that window<sup>1</sup> can select who sould receive the next message he/she sends.

Reviewer: ```suggestion
Each client window has a list with all active users in the top left. The user of that window<sup>1</sup> can select who sould receive the next message he/she sends.
```

> On the top right side one can manage groups. A group saves all currently selected users. The user can see all the groups he is part of. So, by clicking on a group the just described list of usernames will be auto-selected according to the group. **For the same selection of users** two groups can't have the same name.

Reviewer: ```suggestion
On the top right side one can manage groups. A group contains all currently selected users. The user can see all the groups he is part of. So, by clicking on a group the just described list of usernames will be auto-selected according to the group. **For the same selection of users** two groups can't have the same name.
```

> Below all of that is the main part: the actual chat. Here all the messages from all users will be displayed as long as the user was selected to se it (by the sender). That implies that if a new user joins, there won't be any new messages, since no one could send anything to him before.

Reviewer: ```suggestion
Below all of that is the main part: the actual chat. Here all the messages from all users will be displayed as long as the user was selected to see it (by the sender). That implies that if a new user joins, there won't be any new messages, since no one could send anything to him before.
```

Below that there is a simple input field to type in a new message. The message is sent after hitting the return key. 
And on the bottom left there is a button to change the theme for the client between dark-mode and light-mode. 

> Lasty, once a user is done chating and closes the client window, all other clients will be notified and remove that user from their list. If the main window is closed, all client windows will be closed as well and the application will be terminated.

Reviewer: ```suggestion
Lastly, once a user is done chatting and closes the client window, all other clients will be updated and remove that user from their list. If the main window is closed, all client windows will be closed as well and the application will be terminated.
```

<br/>
<sup>1</sup>
The username for each client window is displayed in the title bar of that window and right next to the input field on the bottom.

> # Design

Reviewer: Not there :laughing:

Author: ?

Reviewer: "Design" is not existing

..

# Implementation
..

> # Design Patterns <a name="4"><a/>

Reviewer: note: Observer pattern is used

Reviewer: note: [Fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) pattern is used for all POJOs

...

> # Sources

Reviewer: Google.com :stuck_out_tongue:

> # Advanced Software Enineering Project (DHBW Stuttgart)

Reviewer: ```suggestion
# Advanced Software Engineering Project (DHBW Stuttgart)
```

## A chat simulation

# Content
### 1. [What it does](#1)
### 2. [Design](#design)
### 3. [Implementation](#Implementation)
### 4. [Testing](#testing)
### 5. [Sources](#5)

-----

# What it does <a name="1"><a/>

> This project demonstrates a chat application on one local computer. Once the project is build and started, there is a so called "Main Window". It provides a method for new 'users' to join the chat. After inserting a name which is not yet taken by another user and confirmation, a new window will open, showing the chat client for the new user.

Reviewer: ```suggestion
This project demonstrates a chat application on one local computer. Once the project is built and started, there is a so called "Main Window". It provides a method for new users to join the chat. After inserting a name which is not yet taken by another user and confirmation, a new window will open, showing the chat client for the new user.
```

> Each client window has a list with all active users in the top left. The user of that window<sup>1</sup> can select who sould receive the next message he/she sends.

Reviewer: ```suggestion
Each client window has a list with all active users in the top left. The user of that window<sup>1</sup> can select who should receive the next message he/she sends.
```

> On the top right side one can manage groups. A group contains all currently selected users. The user can see all the groups he is part of. So, by clicking on a group the just described list of usernames will be auto-selected according to the group. **For the same selection of users** two groups can't have the same name.

Reviewer: ```suggestion
On the top right side one can manage groups. A new group contains all currently selected users. The user can see all the groups he is part of. So, by clicking on a group the just described list of usernames will be auto-selected according to the group. **For the same selection of users** two groups can't have the same name.
```

Below all of that is the main part: the actual chat. Here all the messages from all users will be displayed as long as the user was selected to see it (by the sender). That implies that if a new user joins, there won't be any new messages, since no one could send anything to him before.
Below that there is a simple input field to type in a new message. The message is sent after hitting the return key. 

> And on the bottom left there is a button to change the theme for the client between dark-mode and light-mode. 

Reviewer: ```suggestion
And on the bottom left there is a button to change the theme for the client between dark and light mode. 
```

Lastly, once a user is done chatting and closes the client window, all other clients will be updated and remove that user from their list. If the main window is closed, all client windows will be closed as well and the application will be terminated.

<br/>
<sup>1</sup>
The username for each client window is displayed in the title bar of that window and right next to the input field on the bottom.

# Design
The user interface of this application was made in large portions with the Java Swing framework, which is part of the Java Foundation Classes. This framework makes it possible to design and develop responsive program that according to the *Java look and feel* "that looks the same on all platforms". [1]

To further improve the design of the application it was decided to use MiG Layout, an open source layout manager for Java and Swing. It helps to dynamically set the size and position of objects, based on certain rules. This is more reliable, easier to understand and most importantly, it helps avoid bugs in the UI. [2]

> Furthermore, this project uses a few open source image icons to make this application more user-friendly and intuitive. In addition, an option was implemented for each chat participant to change the theme of the user interface between light-mode and dark-mode. All this has no direct effect on the core-functionality and was just implemented to help the user understand and use this application.

Reviewer: ```suggestion
Furthermore, this project uses a few open source image icons to make this application more user friendly and intuitive. In addition, an option was implemented for each chat participant to change the theme of the user interface between light and dark mode. All this has no direct effect on the core functionality and was just implemented to help the user understand and use this application.
```
Bindestriche sind deutsch
